crates/nu_plugin_xpath/src/xpath.rs

```diff
@@ -0,0 +1,166 @@
+use nu_errors::ShellError;
+use nu_protocol::{TaggedDictBuilder, UntaggedValue, Value};
+use nu_source::{Tag, Tagged};
+
+use bigdecimal::{BigDecimal, FromPrimitive};
+
+use sxd_document::parser;
+use sxd_xpath::{Context, Factory};
+
+pub struct Xpath {
+    pub query: String,
+    pub tag: Tag,
+}
+
+impl Xpath {
+    pub fn new() -> Xpath {
+        Xpath {
+            query: String::new(),
+            tag: Tag::unknown(),
+        }
+    }
+}
+
+impl Default for Xpath {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub fn string_to_value(raw: String, query: Tagged<&str>) -> Result<Vec<Value>, ShellError> {
+    execute_xpath_query(raw, query.item.to_string(), query.tag())
+}
+
+fn execute_xpath_query(
+    input_string: String,
+    query_string: String,
+    tag: impl Into<Tag>,
+) -> Result<Vec<Value>, ShellError> {
+    let tag = tag.into();
+    let xpath = build_xpath(&query_string)?;
+
+    let package = parser::parse(&input_string);
+
+    if package.is_err() {
+        return Err(ShellError::labeled_error(
+            "invalid xml document",
+            "invalid xml document",
+            tag.span,
+        ));
+    }
+
+    let package = package.expect("invalid xml document");
+
+    let document = package.as_document();
+    let context = Context::new();
+
+    // leaving this here for augmentation at some point
+    // build_variables(&arguments, &mut context);
+    // build_namespaces(&arguments, &mut context);
+
+    let res = xpath.evaluate(&context, document.root());
+
+    // Some xpath statements can be long, so let's truncate it with ellipsis
+    let mut key = query_string.clone();
```

Comment:
i'm fine with leaving this but we were talking about this in discord today, should the xpath query, which is truncated to 20 chars, be the column label, or should we have a default label of `result` or `value` or something like that?

Reply:
Thought of it but decided to leave it as it was. Up to you!